src/ring_theory/class_group.lean

```diff
@@ -0,0 +1,235 @@
+@[simps]
+noncomputable def fractional_ideal.mk0 [is_dedekind_domain R] :
+  (ideal R)⁰ →* units (fractional_ideal R⁰ K) :=
+{ to_fun := λ I, units.mk0 I ((fractional_ideal.coe_to_fractional_ideal_ne_zero (le_refl R⁰)).mpr
+    (mem_non_zero_divisors_iff_ne_zero.mp I.2)),
+  map_one' := by simp,
+  map_mul' := λ x y, by simp }
+
+/-- Send a nonzero ideal to the corresponding class in the class group. -/
+@[simps]
+noncomputable def class_group.mk0 [is_dedekind_domain R] :
+  (ideal R)⁰ →* class_group R K :=
+(quotient_group.mk' _).comp (fractional_ideal.mk0 K)
+
+variables {K}
+
+lemma quotient_group.mk'_eq_mk' {G : Type*} [group G] {N : subgroup G} [hN : N.normal] {x y : G} :
+  quotient_group.mk' N x = quotient_group.mk' N y ↔ ∃ z ∈ N, x * z = y :=
+(@quotient.eq _ (quotient_group.left_rel _) _ _).trans
+  ⟨λ (h : x⁻¹ * y ∈ N), ⟨_, h, by rw [← mul_assoc, mul_right_inv, one_mul]⟩,
+   λ ⟨z, z_mem, eq_y⟩,
+     by { rw ← eq_y, show x⁻¹ * (x * z) ∈ N, rwa [← mul_assoc, mul_left_inv, one_mul] }⟩
+
+lemma class_group.mk0_eq_mk0_iff_exists_fraction_ring [is_dedekind_domain R] {I J : (ideal R)⁰} :
+  class_group.mk0 K I = class_group.mk0 K J ↔
+    ∃ (x ≠ (0 : K)), span_singleton R⁰ x * I = J :=
+begin
+  simp only [class_group.mk0, monoid_hom.comp_apply, quotient_group.mk'_eq_mk'],
+  split,
+  { rintros ⟨_, ⟨x, rfl⟩, hx⟩,
+    refine ⟨x, x.ne_zero, _⟩,
+    simpa only [mul_comm, coe_mk0, monoid_hom.to_fun_eq_coe, coe_to_principal_ideal, units.coe_mul]
+      using congr_arg (coe : _ → fractional_ideal R⁰ K) hx },
+  { rintros ⟨x, hx, eq_J⟩,
+    refine ⟨_, ⟨units.mk0 x hx, rfl⟩, units.ext _⟩,
+    simpa only [fractional_ideal.mk0_apply, units.coe_mk0, mul_comm, coe_to_principal_ideal,
+        coe_coe, units.coe_mul] using eq_J }
+end
+
+lemma class_group.mk0_eq_mk0_iff [is_dedekind_domain R] {I J : (ideal R)⁰} :
+  class_group.mk0 K I = class_group.mk0 K J ↔
+    ∃ (x y : R) (hx : x ≠ 0) (hy : y ≠ 0), ideal.span {x} * (I : ideal R) = ideal.span {y} * J :=
+begin
+  refine class_group.mk0_eq_mk0_iff_exists_fraction_ring.trans ⟨_, _⟩,
+  { rintros ⟨z, hz, h⟩,
+    refine ⟨(is_localization.sec R⁰ z).1, (is_localization.sec R⁰ z).2.1, _, _, _⟩,
+    { contrapose! hz,
+      rw [← @is_localization.mk'_sec _ _ R⁰ K _ _ _ z, @is_localization.mk'_eq_mul_mk'_one _ _ R⁰ K,
+          mul_eq_zero, hz, ring_hom.map_zero],
+      exact or.inl rfl },
+    { exact mem_non_zero_divisors_iff_ne_zero.mp (is_localization.sec R⁰ z).2.2 },
+    { exact fractional_ideal.coe_ideal_eq_span_singleton_mul_coe_ideal.mp h } },
+  { rintros ⟨x, y, hx, hy, h⟩,
+    have hy' : y ∈ R⁰ := mem_non_zero_divisors_iff_ne_zero.mpr hy,
+    refine ⟨is_localization.mk' K x ⟨y, hy'⟩, _, _⟩,
+    { contrapose! hx,
+      rwa [is_localization.mk'_eq_iff_eq_mul, zero_mul, ← (algebra_map R K).map_zero,
+           (is_fraction_ring.injective R K).eq_iff] at hx },
+    { exact (fractional_ideal.coe_ideal_eq_mk'_mul_coe_ideal K hy').mpr h } },
```

Comment:
I'm not completely happy with the complexity of this proof - mathematically it's just `class_group.mk0_eq_mk0_iff_exists_fraction_ring` + `fractional_ideal.coe_ideal_eq_span_singleton_mul_coe_ideal`, plus some pushing around of arguments.

Anyone up for a bit of golf?